File: .scripts/toggle.rs
// mod main;

use std::{format, process::Command, thread, time::Duration};

fn main() {
    loop {
        let touchpad = get_id("Touchpad");
        let mouse = get_id("Haste  ");

        if mouse != 0 {
            let args = [
                format!("{}", touchpad),
                "Device Enabled".to_string(),
                "0".to_string(),
            ];
            spawn_comand(&args);
            mouse_setup(&mouse);
        } else {
            let args = [
                format!("{}", touchpad),
                "Device Enabled".to_string(),
                "1".to_string(),
            ];
            spawn_comand(&args);
            touchpad_setup(&touchpad);
        }
        thread::sleep(Duration::from_secs(1));
    }
}

fn get_id(name: &str) -> i32 {
    let test = Command::new("xinput").output().unwrap();
    let vec = test.stdout;
    let string1 = String::from_utf8(vec).unwrap();
    let string: Vec<&str> = string1.split('\n').collect();

    let test1 = match string.iter().find(|x| x.contains(name)) {
        Some(x) => x,
        None => return 0,
    };
    let res: Vec<&str> = test1.split('\t').collect();
    let res1: Vec<&str> = res[1].split('=').collect();
    match res1[1].parse::<i32>() {
        Ok(i) => return i,
        Err(_) => return 0,
    }
}

fn kill_multiple_instance(instance: &str) {
    let test = Command::new("pgrep")
        .args(["-fc", instance])
        .output()
        .unwrap();
    // println!("{}", test);
}

fn spawn_comand(args: &[String]) {
    Command::new("xinput")
        .args(["set-prop".to_string()])
        .args(args)
        .output()
        .expect("");
}

fn mouse_setup(mouse: &i32) {
    let args1 = [
        mouse.to_string(),
        "libinput Accel Speed".to_string(),
        "0.25".to_string(),
    ];
    let args2 = [
        mouse.to_string(),
        "libinput Accel Profile Enabled".to_string(),
        "0,".to_string(),
        "1,".to_string(),
        "0".to_string(),
    ];
    spawn_comand(&args1);
    spawn_comand(&args2);
}

fn touchpad_setup(touchpad: &i32) {
    let args1 = [
        touchpad.to_string(),
        "libinput Natural Scrolling Enabled".to_string(),
        "1".to_string(),
    ];
    let args2 = [
        touchpad.to_string(),
        "libinput Tapping Enabled".to_string(),
        "1".to_string(),
    ];
    spawn_comand(&args1);
    spawn_comand(&args2);
}

File: .scripts/toggle.c
#include <X11/X.h>
#include <X11/Xatom.h>
#include <X11/Xlib.h>
#include <X11/extensions/XI.h>
#include <X11/extensions/XI2.h>
#include <X11/extensions/XInput.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

#define KEYBOARD 1
#define POINTER 2

XDevice *getDevice(Display *display, char *device_name, int num_class);
void setUpTouchpad(Display *display, XDevice *touchpad, float *value);
void changeProperty(Display *display, XDevice *device, char *property,
                    float *value);
void mainLoop(XDevice *mouse, XDevice *touchpad, Display *display,
              float *value);

int main() {
  Display *display;
  XDevice *mouse, *touchpad;
  float value[4] = {};

  display = XOpenDisplay(NULL);
  if (display == NULL) {
    fprintf(stderr, "Unable to open display.\n");
    exit(1);
  }
  touchpad = getDevice(display, "VEN_04F3:00 04F3:32AA Touchpad", POINTER);
  if (touchpad == NULL) {
    fprintf(stderr, "Unable to open touchpad.\n");
    exit(1);
  }
  setUpTouchpad(display, touchpad, value);

  while (1) {
    mainLoop(mouse, touchpad, display, value);
    sleep(1);
  }

  XFlush(display);
  return 0;
}

void setUpTouchpad(Display *display, XDevice *touchpad, float *value) {
  Atom accelSpeedAtom = XInternAtom(display, "Device Enabled", False);
  Atom type;
  int format;
  unsigned char *data = NULL;
  unsigned long nitems, bytes_after;

  XGetDeviceProperty(display, touchpad, accelSpeedAtom, 0, 1000, False,
                     AnyPropertyType, &type, &format, &nitems, &bytes_after,
                     &data);
  if (data[0] == 1) {
    value[0] = 1;
    changeProperty(display, touchpad, "libinput Natural Scrolling Enabled",
                   value);
    changeProperty(display, touchpad, "libinput Tapping Enabled", value);
  }
  free(data);
}

void mainLoop(XDevice *mouse, XDevice *touchpad, Display *display,
              float *value) {
  mouse = getDevice(display, "Kingston HyperX Pulsefire Haste", POINTER);
  if (mouse != NULL) {
    value[0] = 0.25;
    changeProperty(display, mouse, "libinput Accel Speed", value);
    value[0] = 0;
    value[1] = 1;
    value[2] = 0;
    changeProperty(display, mouse, "libinput Accel Profile Enabled", value);
    XCloseDevice(display, mouse);
    value[0] = 0;
  } else {
    value[0] = 1;
  }
  changeProperty(display, touchpad, "Device Enabled", value);
}

XDevice *getDevice(Display *display, char *device_name, int num_class) {
  XDeviceInfo *devices;
  int num_devices;

  devices = XListInputDevices(display, &num_devices);
  if (devices == NULL) {
    fprintf(stderr, "Unable to query devices.\n");
    exit(1);
  }

  int i;
  for (i = 0; i < num_devices; ++i) {
    if (strcmp(devices[i].name, device_name) == 0 &&
        devices[i].num_classes == num_class) {
      // Open the device
      XDevice *device = XOpenDevice(display, devices[i].id);
      if (device == NULL) {
        fprintf(stderr, "Unable to open device.\n");
        exit(1);
      }
      free(devices);
      return device;
    }
  }
  free(devices);
  return NULL;
}

void changeProperty(Display *display, XDevice *device, char *property,
                    float *value) {
  int num_devices;
  Atom prop = None, type;
  int format, skip = 1;
  unsigned char *data = NULL;
  unsigned long nitems, bytes_after;

  Atom accelSpeedAtom = XInternAtom(display, property, False);

  XGetDeviceProperty(display, device, accelSpeedAtom, 0, 1000, False,
                     AnyPropertyType, &type, &format, &nitems, &bytes_after,
                     &data);

  if (type == XA_INTEGER) {
    for (int i = 0; i < nitems; i++) {
      if (data[i] != (int)value[i]) {
        data[i] = (int)value[i];
        skip = 0;
      }
    }
  } else {
    for (int i = 0; i < nitems; i++) {
      if (((float *)data)[i] != value[i]) {
        ((float *)data)[i] = value[i];
        skip = 0;
      }
    }
  }

  if (!skip) {
    XChangeDeviceProperty(display, device, accelSpeedAtom, type, format,
                          PropModeReplace, data, nitems);
  }

  free(data);
}
